ngraph/core/shape_inference/include/gather_shape_inference.hpp

```diff
@@ -0,0 +1,122 @@
+//
+#pragma once
+#include <openvino/op/util/gather_base.hpp>
+
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace util {
+
+template <class RankT, class ShapeT>
+void inline create_shape(RankT& rank, ShapeT& shape) {
+    return;
+}
+
+template <class RankT>
+void inline create_shape(RankT& rank, PartialShape& shape) {
+    shape = PartialShape::dynamic(rank);
+}
+
+template <class ShapeT, typename = typename std::enable_if<!std::is_same<ShapeT, PartialShape>::value>::type>
+void inline create_shape(int64_t& rank, ShapeT& shape) {
+    shape.resize(rank);
+}
+
+template <class T>
+void shape_infer(GatherBase* op,
+                 const std::vector<T>& input_shapes,
+                 std::vector<T>& output_shapes,
+                 const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data = {}) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 3 && output_shapes.size() == 1);
+    const auto& data_pshape = input_shapes[0];
+    const auto& indices_pshape = input_shapes[1];
+    const auto& axis_pshape = input_shapes[2];
+    auto data_rank = data_pshape.rank();
+    auto indices_rank = indices_pshape.rank();
+    auto axis_rank = axis_pshape.rank();
+    auto& output_pshape = output_shapes[0];
+
+    if (axis_rank.is_static() && axis_pshape.is_static()) {
+        const auto axis_is_scalar = axis_rank.get_length() == 0;
+        const auto axis_has_one_elem = axis_rank.get_length() == 1 && axis_pshape[0].get_length() == 1;
+        NODE_VALIDATION_CHECK(op,
+                              axis_is_scalar || axis_has_one_elem,
+                              "Axis input must be scalar or have 1 element. But instead got axis_shape = ",
+                              axis_pshape);
+    }
+    int64_t batch_dims = op->get_batch_dims();
+    if (indices_rank.is_static()) {
+        const auto indices_rank_value = indices_rank.get_length();
+        batch_dims = ov::normalize_axis(op, batch_dims, indices_rank_value, -indices_rank_value, indices_rank_value);
+    }
+
+    std::vector<int64_t> axes_val;
+    bool axis_is_set = get_data_as_int64<T>(2, op, axes_val, constant_data);
+    int64_t axis = 0;
+
+    if (axis_is_set) {
+        axis = axes_val[0];
+    }
+
+    if (axis_is_set && data_rank.is_static()) {
+        axis = ov::normalize_axis(op, axis, data_rank);
+    }
+
+    if (axis_is_set) {
+        // batch_dims, axis both can be positive by default or after normalization if data_rank &
+        // indices_rank are static.
+        // If at least one of them is negative we cannot check their consistency.
+        NODE_VALIDATION_CHECK(op,
+                              batch_dims <= axis || batch_dims < 0 || axis < 0,
+                              "After normalization batch_dims must be <= axis. But instead got: batch_dims = ",
+                              batch_dims,
+                              ", axis = ",
+                              axis);
+    }
```

Comment:
This is more readable. Also -- there is no need to require batch_dims to be normalized. I can imagine that freshly read model could have dynamic rank from the data input and negative axis. In this case current code will throw an exception. But instead it could propagate dynamic shape. 
```suggestion
    if (axis_is_set) {
        axis = axes_val[0];

        if (data_rank.is_static())
            axis = ov::normalize_axis(op, axis, data_rank);

        // batch_dims, axis both can be positive by default or after normalization if data_rank &
        // indices_rank are static.
        // If at least one of them is negative we cannot check their consistency.
        NODE_VALIDATION_CHECK(op,
                          batch_dims <= axis || batch_dims < 0 || axis < 0,
                          "After normalization batch_dims must be <= axis. But instead got: batch_dims = ",
                          batch_dims,
                          ", axis = ",
                          axis);
```